File: src/main/python/com/hazza/leetcode_by_python/SegmentTwo/No63_Unique_Paths_II.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/6/13 20:09
# https://leetcode.com/problems/unique-paths-ii/
from typing import List


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        dp = [0] * (n + 1)

        dp[1] = 1
        for i in range(m):
            for j in range(n):
                if obstacleGrid[i][j] == 1:
                    dp[j + 1] = 0
                else:
                    dp[j + 1] = dp[j + 1] + dp[j]

        return dp[n]


if __name__ == '__main__':
    solution = Solution()

    input = [
        [0, 0, 0],
        [0, 1, 0],
        [0, 0, 0]
    ]
    assert solution.uniquePathsWithObstacles(input) == 2

File: src/main/java/com/hazza/leetcode/SegmentTwo/No63UniquePathsII.java
package com.hazza.leetcode.SegmentTwo;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/unique-paths-ii/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-18
 * Time: 7:08 PM
 */
public class No63UniquePathsII {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int row = obstacleGrid.length;
        int col = obstacleGrid[0].length;
        int[] dp = new int[col + 1];
        dp[1] = 1;
        for (int i = 0; i < row; i++) {
            for (int j = 1; j <= col; j++) {
                if (obstacleGrid[i][j-1] == 1) dp[j] = 0;
                else dp[j] = dp[j] + dp[j - 1];
            }
        }

        return dp[col];
    }
}
